### ISSUE TITLE
idea: swap arguments and support curry

### ISSUE THREAD

**Issue by hex13** [mweststrate/immer#36]
(related to #24 )

idea:
1. swap parameters (recipe, state) instead of other way around. It could be counterintuitive but it could help with partial application(or curry) and reusing transformations.
2. if `immer` is called with only first argument (recipe function) it would not transform but return other function with recipe function bound 
3.  allow for passing additional parameters to `recipe` function

WHY?
 it would help reuse transformation, treat them as objects. For example when using with Redux  #31 or with `map` support #35 

```javascript
const reducer = immer((state,  action) => {
    switch (action.type) {
        case 'add':
             state.arr.push(action.something);
            break;
    }
});

createStore(reducer, initialState);
```

**Comment by lukasbuenger** [comment]
While I absolutely agree with all of your points, I think it really does look counterintuitive to anybody not embracing FP style and I'm not quite sure whether this library wants to enforce that. 
In this case though, with the immer being stateless and of a fixed arity 2, it is quite easy to decorate `immer` somewhere in your project and just import from there.
```js
import immer from 'immer'

export default (action, state) => immer(state, action)
```

If ramda is your thing you could even just put this somewhere ...
```js
import immer from 'immer'
import { flip, uncurryN } from 'ramda'

export default uncurryN(2, flip(immer))
```
...and you'd get a curried version of `immer` with signature: `action -> state -> state`

**Comment by mweststrate** [comment]
I quite like the idea. I'm gonna look closer next week, but since the arguments are pretty distinguishable, I think all these could be supported

```
/* 1 */ immer(state, fn) // default, most friendly style
/* 2 */ immer(fn)(state, ...args) // curried style
/* 3 */ immer(fn, state) // would this style have any merits? Otherwise probably better left out
```

**Comment by lukasbuenger** [comment]
Given the fact that signatures 2 and 3 are probably preferred by FP people only, I think it is safe to say that FP minded users groks the concept if you put something like the following in the docs. 
To support the behavior by default you'll have to either maintain at least some dependencies or implement curry yourself (again). 

Using ramda:
```js
import immer from 'immer'
import { flip, uncurryN } from 'ramda' 

export default uncurryN(2, flip(immer))
```

Using lodash:
```js
import immer from 'immer'
import curry from 'lodash.curry'

export default curry(
  (action, state) => immer(state, action)
)
```

**Comment by mweststrate** [comment]
Implemented in 0.4.1, see https://github.com/mweststrate/immer#currying
